flowmachine/flowmachine/features/subscriber/mobility_classification.py

```diff
@@ -0,0 +1,124 @@
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from typing import List
+
+from flowmachine.features.subscriber.metaclasses import SubscriberFeature
+from flowmachine.features.utilities.subscriber_locations import BaseLocation
+from flowmachine.core.errors import InvalidSpatialUnitError
+
+
+class MobilityClassification(SubscriberFeature):
+    """
+    Based on subscribers' reference locations in a sequence of reference
+    periods, classify each subscriber as having one of the following mobility
+    types (the assigned label corresponds to the first of these criteria that
+    is true for a given subscriber):
+
+    - 'unlocated': Subscriber has a NULL location in the most recent period
+    - 'irregular': Subscriber is not active in at least one of the reference
+      periods
+    - 'not_always_locatable': Subscriber has a NULL location in at least one of
+      the reference periods
+    - 'mobile': Subscriber spent fewer than 'stay_length_threshold' consecutive
+      periods at any single location
+    - 'stable': Subscriber spent at least 'stay_length_threshold' consecutive
+      periods at the same location
+
+    Only subscribers appearing in the result of the reference location query
+    for the most recent period are included in the result of this query (i.e.
+    subscribers absent from the query result can be assumed to fall into a
+    sixth category: "not active in the most recent period").
+
+    Parameters
+    ----------
+    locations : list of BaseLocation
+        List of reference location queries, each returning a single location
+        per subscriber (or NULL location for subscribers that are active but
+        unlocatable). The list is assumed to be sorted into ascending
+        chronological order.
+    stay_length_threshold : int, default 3
+        Minimum number of consecutive periods over which a subscriber's
+        location must remain the same for that subscriber to be classified as
+        'stable'.
+    """
+
+    def __init__(self, locations: List[BaseLocation], stay_length_threshold: int = 3):
+        self.locations = locations
+        if len(set(l.spatial_unit for l in self.locations)) > 1:
+            raise InvalidSpatialUnitError(
+                "MobilityClassification requires all input locations to have the same spatial unit"
+            )
+        self.stay_length_threshold = int(stay_length_threshold)
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return ["subscriber", "value"]
+
+    def _make_query(self) -> str:
+        loc_cols_string = ", ".join(self.locations[0].spatial_unit.location_id_columns)
+
+        # Workaround to count rows where all location columns are null, if there are multiple location columns
+        if len(self.locations[0].spatial_unit.location_id_columns) > 1:
+            count_unlocatable = f"sum((NOT (({loc_cols_string}) IS NULL))::int)"
```

Comment:
This is actually `count_locatable` rather than _un_ locatable, right?

Reply:
Might be more readable to use `num_nulls()`?

Reply:
Cos then you can do:

`sum(num_nulls({loc_cols_string}))` for both cases, and check for above zero below.

Reply:
Yes, you're right, this is actually `count_locatable`.

I'm not sure `num_nulls` helps to make this cleaner - I think we want to count the rows where _all_ of the location columns are null, which would mean counting the rows with `num_nulls({loc_cols_string}) = <number_of_location_cols>`. (I don't think there's any situation where the location columns in one row might have a mix of null and non-null values, but if that ever did occur I expect we'd want to treat that as a known location rather than "unlocatable").

Perhaps flipping this round to `count_unlocatable = 0` instead of `count_locatable < {len(self.locations)}` is a better approach. We could then do `bool_or(({loc_cols_string}) IS NULL) AS sometimes_unlocatable`.

Reply:
"location columns in one row might have a mix of null and non-null values" cell with longitude or latitude missing?

Reply:
Hmm, I hadn't considered that. I was thinking of 'unlocatable' in the sense used in MajorityLocation - i.e. "we couldn't assign this subscriber to a single location". Whereas a cell with missing longitude/latitude would be "we were able to assign this subscriber to a single cell, but we don't know where that cell _is_".

I think if we're using this query with versioned-cell-level locations, the first definition makes more sense - the subscriber is still locatable to a specific versioned cell. If we wanted to exclude cells with unknown locations, we'd be better off excluding those earlier so that they don't appear in the outputs of the 'locations' sub-queries.

Reply:
Yeah, I can see your point. I too have difficulty conjuring a case where we'd be only partially assigning to be honest.